### ISSUE TITLE
Declaration file generated reports error on compilation while there were no errors reported during generation

### ISSUE THREAD

**Issue by sheetalkamat** [Microsoft/TypeScript#5659]
Test case: https://github.com/Microsoft/TypeScript/blob/master/tests/cases/conformance/types/stringLiteral/stringLiteralTypesAsTypeParameterConstraint01.ts#L11

Repro:
Compile a.ts with --d
No errors reported
Compile a.d.ts generated, reports error

a.ts:
```typescript
function foo<T extends "foo">(f: (x: T) => T) {
    return f;
}

function bar<T extends "foo" | "bar">(f: (x: T) => T) {
    return f;
}

let f = foo(x => x);
let fResult = f("foo");

let g = foo((x => x));
let gResult = g("foo");

let h = bar(x => x);
let hResult = h("foo");
hResult = h("bar");
```

**Comment by DanielRosenwasser** [comment]
Oof. Specifically, the errors are:

```shell
$ tsc .\c.d.ts
c.d.ts(3,16): error TS2382: Specialized overload signature is not assignable to any non-specialized signature.
c.d.ts(5,16): error TS2382: Specialized overload signature is not assignable to any non-specialized signature.
```

Luckily we're removing restrictions on specialized signatures.

**Comment by DanielRosenwasser** [comment]
Actually, something is making this consider itself a specialized signature, so I'm going to fix this first.

**Comment by DanielRosenwasser** [comment]
This is due to #6143. The error is on `f` and `g`, not on `foo` and `bar`.
